src/decorator.c

```diff
@@ -0,0 +1,399 @@
+    lua_pushnil(L);    // obj; code; UiName; name; Ui; name; nil
+    lua_rawset(L, -3); // obj; code; UiName; name; Ui;
+    lua_pop(L, 1);
+
+    // remove info in table UiName
+    // obj; code; UiName; name
+    lua_pushnil(L);    // obj; code; UiName; name; nil
+    lua_rawset(L, -3); // obj; code; UiName;
+    lua_pop(L, 1);
+  }
+  // obj; code
+
+  // remove timer
+  lua_pushstring(L, LUAVGL_BUILTIN_STR_UserData);
+  lua_rawget(L, LUA_REGISTRYINDEX); // obj; code; UserData
+
+  lua_pushvalue(L, 1);
+  lua_rawget(L, -2); // obj; code; UserData; objData
+
+  lua_pushcfunction(L, luavgl_timer_delete);
+
+  lua_pushstring(L, "__timer");
+  lua_gettable(L, -2);
+  // obj; code; UserData; objData; luavgl_timer_delete; objData["__timer"]
+  if (!lua_isnil(L, -1)) {
+    lua_call(L, 1, 0);
+    lua_pop(L, 2);
+  } else {
+    lua_pop(L, 4);
+  }
+  // obj; code
+  return 0;
+}
+
+static void luavgl_decorator_ondelete(lua_State *L)
+{
+  // key; tb; stb; itb; obj
+  lua_pushvalue(L, -1);
+  // obj
+
+  lua_pushcfunction(L, luavgl_obj_on_event);
+  lua_pushvalue(L, -2);
+  lua_pushinteger(L, LV_EVENT_DELETE);
+  lua_pushcfunction(L, luavgl_decorator_obj_delete_event);
+  lua_call(L, 3, 0);
+  // obj
+
+  lua_pop(L, 1);
+  // key; tb; stb; itb; obj
+}
+
+static int luavgl_decorator_obj(lua_State *L)
+{
+
+  if (!lua_istable(L, 2)) {
+    luaL_argerror(L, 2, "ui wrapper args 2 should be table.");
+    return 0;
+  }
+
+  // key; tb
+  luavgl_decorator_splite_table(L); // 需要分离出 prebuild 和 afterbuild
+  // key; tb; stb; itb
+  luavgl_decorator_create_obj(L);
+  // key; tb; stb; itb; obj
+
+  // call "prebuild"
+  lua_pushstring(L, "prebuild");
+  lua_rawget(L, LUA_REGISTRYINDEX);
+  if (lua_isnil(L, -1)) {
+    lua_pop(L, 1);
+  } else {
+    lua_pushvalue(L, -2);
+    lua_call(L, 1, 0);
+
+    lua_pushstring(L, "prebuild");
+    lua_pushnil(L);
+    lua_rawset(L, LUA_REGISTRYINDEX);
+  }
+
+  // Call each function keeping the following stack state
+  // key; tb; stb; itb; obj
+  luavgl_decorator_bind_child(L);
+  luavgl_decorator_ondelete(L);
+  luavgl_decorator_decorate(L);
+
+  // call "afterbuild"
+  lua_pushstring(L, "afterbuild");
+  lua_rawget(L, LUA_REGISTRYINDEX);
+  if (lua_isnil(L, -1)) {
+    lua_pop(L, 1);
+  } else {
+    lua_pushvalue(L, -2);
+    lua_call(L, 1, 0);
+
+    lua_pushstring(L, "afterbuild");
+    lua_pushnil(L);
+    lua_rawset(L, LUA_REGISTRYINDEX);
+  }
+  return 1;
+}
+
+// Depending on the "widget_create_methods" array for initialization
+extern const luaL_Reg widget_create_methods[];
+extern int luavgl_style_create(lua_State *L);
+
+static void luavgl_decorator_init(lua_State *L)
+{
+  lua_newtable(L);
+  lua_setfield(L, LUA_REGISTRYINDEX, LUAVGL_BUILTIN_STR_Ui);
+
+  lua_newtable(L);
+  lua_setfield(L, LUA_REGISTRYINDEX, LUAVGL_BUILTIN_STR_UiName);
+
+  lua_newtable(L);
+  lua_setfield(L, LUA_REGISTRYINDEX, LUAVGL_BUILTIN_STR_UserData);
+
+  lua_register(L, "GetUi", _luavgl_decorator_GetUi);
```

Comment:
I think since luavgl is used, the main function is to provide convenient ui creation, so for some common functions you can consider not putting them under the lvgl table.